src/config/compiler.rs

```diff
@@ -1,9 +1,45 @@
-use indexmap::IndexSet;
+// Workaround for a false positive. The function `create_data_dir_lock` cannot be const.
+#![allow(clippy::missing_const_for_fn)]
 
 use super::{
     builder::ConfigBuilder, graph::Graph, id::Inputs, transform::get_transform_output_ids,
     validation, Config, OutputId,
 };
+use std::fs::File;
+use std::path::PathBuf;
+
+use indexmap::IndexSet;
+
+#[cfg(not(test))]
+use fs4::FileExt;
+
+#[cfg(not(test))]
+fn create_data_dir_lock(data_dir: &Option<PathBuf>) -> Result<Option<File>, String> {
+    if let Some(data_dir) = data_dir {
+        match acquire_exclusive_lock(data_dir) {
+            Ok(lock) => Ok(Some(lock)),
+            Err(e) => Err(e),
+        }
+    } else {
+        Ok(None)
+    }
+}
+
+#[cfg(test)]
+fn create_data_dir_lock(_data_dir: &Option<PathBuf>) -> Result<Option<File>, String> {
+    // We should make all tests work with unique directories but this requires extensive changes.
+    Ok(None)
+}
+
+#[cfg(not(test))]
+fn acquire_exclusive_lock(path: &std::path::Path) -> Result<std::fs::File, String> {
+    let lock_path = path.join(".lock");
+    let lock = std::fs::File::create(&lock_path).map_err(|e| e.to_string())?;
+    match lock.try_lock_exclusive() {
+        Ok(()) => Ok(lock),
+        Err(e) => Err(format!("Couldn't lock {lock_path:?}. Error: {e}")),
+    }
+}
```

Comment:
Feels like you could just move this logic into `create_data_dir_lock` since nothing else uses it.